docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+  by calling `Status(n)` remotely on all peers *n*.
+  
+- `CreateRequest`: regularly checks whether certain blocks have no
+  open request. If a block does not have an open request, it requests
+  one from a peer. It does so by calling `Block(n,h)` remotely on one
+  peer *n* for a missing height *h*.
+  
+> We have left the strategy how peers are selected unspecified, and
+> the currently existing different implementations of Fastsync differ
+> in this aspect. 
+
+The functions `Status` and `Block` are called by asynchronous
+RPC. When they return, the following functions are called:
+
+- `OnStatusResponse(addr Address, height int64)`: The full node with
+  address *addr* returns its current height. The function updates the height
+  information about *addr*, and may also increase *TargetHeight*.
+  
+- `OnBlockResponse(addr Address, b Block)`. The full node with
+  address *addr* returns a block. It is added to *blockstore*. Then
+  the auxiliary function `Execute` is called.
+
+- `Execute()`: Iterates over the *blockstore*.  Checks soundness of
+  the blocks, and
+  executes the transactions of a sound block and updates *state*. 
+
+> In addition to the functions above, the following two features are
+> implemented in Fastsync V2
+
+#### **[FS-V2-PEER-REMOVE]**:
+Periodically, *peerTimeStamp* and *peerRate* are analyzed. If a peer *p*
+has not provided a block recently (check of *peerTimeStamp[p]*) or it
+has not provided sufficiently many data (check of *peerRate[p]*), then
+*p* is removed from *peerIDs*.
+  
+#### **[FS-V2-TIMEOUT]**:
+
+*Fastsync V2* starts a timeout whenever a block is
+executed (that is, when the height is incremented). If the timeout expires
+before the next block is executed, *Fastsync* terminates.
+We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
+with failure, otherwise it terminates successfully.
+
+
+
+
+### Details
+
+<!--
+> Function signatures followed by pseudocode (optional) and a list of features (required):
+> - Implementation remarks (optional)
+>   - e.g. (local/remote) function called in the body of this function
+> - Expected precondition
+> - Expected postcondition
+> - Error condition
+---->
+
+
+
+```go
+func QueryStatus()
+```
+- Expected precondition
+    - peerIDs initialized and non-empty
+- Expected postcondition
+    - call asynchronously `Status(n)` at each peer *n* in *peerIDs*.
+- Error condition
+    - fails if precondition is violated
+----
+
+```go
+func OnStatusResponse(addr Address, ht int64)
+```
+- Comment
+    - *ht* is a height
+	- peers can provide the status without being called
+- Expected precondition
+    - *peerHeights(addr) <= ht*
+- Expected postcondition
+    - *peerHeights(addr) = ht*
+	- *TargetHeight* is updated
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs* (that is,
+      *addr* is removed from *peerIDs*)
+- Timeout condition
+    - if `OnStatusResponse(addr, ht)` was not invoked within *2 Delta* after
```

Comment:
Note that in V2 this is not the case, i.e., it is not tracked when statusResponse arrived. Therefore, peers are not removed upon late statusReply.